File: java/Algorithms/test/Collections/MaxPriorityQueueTest.java
package Collections;

import Sorts.HeapSort;
import Sorts.Point;
import Sorts.Shuffle;
import Sorts.SortOrder;
import org.junit.Test;

import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertThat;

public class MaxPriorityQueueTest {
    @Test
    public void testMaxPriorityQueue_InitWithKey() throws Exception {
        Integer[] a = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
        Shuffle.shuffle(a);

        MaxPriorityQueue<Integer> pq = new MaxPriorityQueue<Integer>(a);
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));

        for (int i = 0; i < 10; i++) {
            Integer result = pq.max();
            assertThat(result, equalTo(10 - i - 1));
            result = pq.delMax();
            assertThat(result, equalTo(10 - i - 1));
        }
        assertThat(pq.isEmpty(), equalTo(true));
        assertThat(pq.size(), equalTo(0));
    }

    @Test
    public void testMaxPriorityQueue_Insert() throws Exception {
        Integer[] a = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
        Shuffle.shuffle(a);

        MaxPriorityQueue<Integer> pq = new MaxPriorityQueue<Integer>();
        for (Integer value : a) {
            pq.insert(value);
        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));

        for (int i = 0; i < 10; i++) {
            Integer result = pq.max();
            assertThat(result, equalTo(10 - i - 1));
            result = pq.delMax();
            assertThat(result, equalTo(10 - i - 1));
        }
        assertThat(pq.isEmpty(), equalTo(true));
        assertThat(pq.size(), equalTo(0));
    }

    @Test
    public void testMaxPriorityQueue_Iterator() throws Exception {
        Integer[] a = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
        Shuffle.shuffle(a);

        MaxPriorityQueue<Integer> pq = new MaxPriorityQueue<Integer>();
        for (Integer value : a) {
            pq.insert(value);
        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));

        Integer expectValue = 9;
        for (Integer value : pq) {
            assertThat(value, equalTo(expectValue--));
        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));
    }

    @Test
    public void testMaxPriorityQueue_Comparator() throws Exception {
        Point[] points = new Point[10];
        for (int i = 0; i < 10; i++) {
            points[i] = new Point(i, 10 - i);
        }
        Shuffle.shuffle(points);

        MaxPriorityQueue<Point> pq = new MaxPriorityQueue<Point>(Point.X_ORDER);
        for (Point value : points) {
            pq.insert(value);
        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));

        Point[] expectPoints = new Point[10];
        for (int i = 0; i < 10; i++) {
            expectPoints[i] = new Point(i, 10 - i);
        }
        Shuffle.shuffle(expectPoints);
        HeapSort.sort(expectPoints, Point.X_ORDER, SortOrder.DESC);

        for (int i = 0; i < 10; i++) {
            Point result = pq.max();
            assertThat(result.getX(), equalTo(expectPoints[i].getX()));
            assertThat(result.getY(), equalTo(expectPoints[i].getY()));
            result = pq.delMax();
            assertThat(result.getX(), equalTo(expectPoints[i].getX()));
            assertThat(result.getY(), equalTo(expectPoints[i].getY()));
        }
        assertThat(pq.isEmpty(), equalTo(true));
        assertThat(pq.size(), equalTo(0));
    }

    @Test
    public void testMaxPriorityQueue_Comparator_Iterator() throws Exception {
        Point[] points = new Point[10];
        for (int i = 0; i < 10; i++) {
            points[i] = new Point(i, 10 - i);
        }
        Shuffle.shuffle(points);

        MaxPriorityQueue<Point> pq = new MaxPriorityQueue<Point>(Point.X_ORDER);
        for (Point value : points) {
            pq.insert(value);
        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));

        Point[] expectPoints = new Point[10];
        for (int i = 0; i < 10; i++) {
            expectPoints[i] = new Point(i, 10 - i);
        }
        Shuffle.shuffle(expectPoints);
        HeapSort.sort(expectPoints, Point.X_ORDER, SortOrder.DESC);

        Integer expectIndex = 0;
        for (Point value : pq) {
            assertThat(value.getX(), equalTo(expectPoints[expectIndex].getX()));
            assertThat(value.getY(), equalTo(expectPoints[expectIndex].getY()));
            expectIndex++;

        }
        assertThat(pq.isEmpty(), equalTo(false));
        assertThat(pq.size(), equalTo(10));
    }
}

File: c#/BigEgg.Algorithm/BigEgg.Algorithm.Test/Collections/MaxPriorityQueueTest.cs
using BigEgg.Algorithm.Collections;
using BigEgg.Algorithm.Sorts;
using BigEgg.Algorithm.Test.Sorts;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace BigEgg.Algorithm.Test.Collections
{
    [TestClass]
    public class MaxPriorityQueueTest
    {
        [TestMethod]
        public void MaxPriorityQueueTest_InitWithKey()
        {
            int[] a = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
            Shuffle.Do(a);

            MaxPriorityQueue<int> pq = new MaxPriorityQueue<int>(a);
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());

            for (int i = 0; i < 10; i++)
            {
                int result = pq.Max();
                Assert.AreEqual(10 - i - 1, result);
                result = pq.DelMax();
                Assert.AreEqual(10 - i - 1, result);
            }
            Assert.IsTrue(pq.IsEmpty());
            Assert.AreEqual(0, pq.Size());
        }

        [TestMethod]
        public void MaxPriorityQueueTest_Insert()
        {
            int[] a = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
            Shuffle.Do(a);

            MaxPriorityQueue<int> pq = new MaxPriorityQueue<int>();
            foreach (var value in a)
            {
                pq.Insert(value);
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());

            for (int i = 0; i < 10; i++)
            {
                int result = pq.Max();
                Assert.AreEqual(10 - i - 1, result);
                result = pq.DelMax();
                Assert.AreEqual(10 - i - 1, result);
            }
            Assert.IsTrue(pq.IsEmpty());
            Assert.AreEqual(0, pq.Size());
        }

        [TestMethod]
        public void MaxPriorityQueueTest_Enumerator()
        {
            int[] a = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
            Shuffle.Do(a);

            MaxPriorityQueue<int> pq = new MaxPriorityQueue<int>();
            foreach (var value in a)
            {
                pq.Insert(value);
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());

            int expectedValue = 9;
            foreach (var value in pq)
            {
                Assert.AreEqual(expectedValue--, value);
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());
        }

        [TestMethod]
        public void MaxPriorityQueueTest_Comparer()
        {
            Point[] points = new Point[10];
            for (int i = 0; i < 10; i++)
            {
                points[i] = new Point(i, 10 - i);
            }
            Shuffle.Do(points);

            MaxPriorityQueue<Point> pq = new MaxPriorityQueue<Point>(Point.X_ORDER);
            foreach (var value in points)
            {
                pq.Insert(value);
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());

            Point[] expectPoints = new Point[10];
            for (int i = 0; i < 10; i++)
            {
                expectPoints[i] = new Point(i, 10 - i);
            }
            Shuffle.Do(expectPoints);
            HeapSort.Sort(expectPoints, Point.X_ORDER, SortOrder.DESC);

            for (int i = 0; i < 10; i++)
            {
                var result = pq.Max();
                Assert.AreEqual(expectPoints[i].X, result.X);
                Assert.AreEqual(expectPoints[i].Y, result.Y);
                result = pq.DelMax();
                Assert.AreEqual(expectPoints[i].X, result.X);
                Assert.AreEqual(expectPoints[i].Y, result.Y);
            }
            Assert.IsTrue(pq.IsEmpty());
            Assert.AreEqual(0, pq.Size());
        }

        [TestMethod]
        public void MaxPriorityQueueTest_Comparer_Enumerator()
        {
            Point[] points = new Point[10];
            for (int i = 0; i < 10; i++)
            {
                points[i] = new Point(i, 10 - i);
            }
            Shuffle.Do(points);

            MaxPriorityQueue<Point> pq = new MaxPriorityQueue<Point>(Point.X_ORDER);
            foreach (var value in points)
            {
                pq.Insert(value);
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());

            Point[] expectPoints = new Point[10];
            for (int i = 0; i < 10; i++)
            {
                expectPoints[i] = new Point(i, 10 - i);
            }
            Shuffle.Do(expectPoints);
            HeapSort.Sort(expectPoints, Point.X_ORDER, SortOrder.DESC);

            int expectedIndex = 0;
            foreach (var result in pq)
            {
                Assert.AreEqual(expectPoints[expectedIndex].X, result.X);
                Assert.AreEqual(expectPoints[expectedIndex].Y, result.Y);
                expectedIndex++;
            }
            Assert.IsFalse(pq.IsEmpty());
            Assert.AreEqual(10, pq.Size());
        }
    }
}
